### PR TITLE
Move CorsHandler to server

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently we duplicate our specialized cors logic in all transport
plugins. This is unnecessary as it could be implemented in a single 
place. This commit moves the logic to server. Additionally it fixes a
but where we are incorrectly closing http channels on early Cors
responses.

### CODE REVIEW

**Comment by tbrooks8** [comment]
This deletes the various handlers and tests and moves the code to a single handler (CorsHandler) and tests (CorsHandlerTests). Additionally, it adds a bit of test coverage in DefaultRestChannelTests.

The only behavioral changes is no longer closing channels after early Cors responses (bug) and moving Cors after pipelining so that pipelining is correctly observed.

File: server/src/main/java/org/elasticsearch/http/CorsHandler.java [hunk 1]
```diff
@@ -71,12 +78,175 @@
...
+        if (config.isCredentialsAllowed()) {
+            response.addHeader(ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+        }
+    }
+
+    private void setMaxAge(final HttpResponse response) {
+        response.addHeader(ACCESS_CONTROL_MAX_AGE, Long.toString(config.maxAge));
+    }
+
+
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/http/CorsHandler.java:249]: NIT: extra empty line

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequestHandler.java [hunk 2]
```diff
@@ -23,23 +23,40 @@
...
+        this.corsHandler = corsHandler;
     }
 
     @Override
     protected void channelRead0(ChannelHandlerContext ctx, HttpPipelinedRequest httpRequest) {
         final Netty4HttpChannel channel = ctx.channel().attr(Netty4HttpServerTransport.HTTP_CHANNEL_KEY).get();
         boolean success = false;
         try {
-            serverTransport.incomingRequest(httpRequest, channel);
+            HttpResponse earlyResponse = corsHandler.handleInbound(httpRequest);
```
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequestHandler.java:53]: Maybe we could just move this whole logic into `org.elasticsearch.http.AbstractHttpServerTransport#incomingRequest` to make things even drier?

**Review by original-brownbear — COMMENTED** [review]
One question on making this even drier

File: plugins/transport-nio/src/main/java/org/elasticsearch/http/nio/HttpReadWriteHandler.java [hunk 3]
```diff
@@ -150,16 +145,15 @@ public void close() throws IOException {
         }
     }
 
-    @SuppressWarnings("unchecked")
     private void handleRequest(Object msg) {
-        final HttpPipelinedRequest pipelinedRequest = (HttpPipelinedRequest) msg;
+        final HttpPipelinedRequest httpRequest = (HttpPipelinedRequest) msg;
```
Comment (original-brownbear) [inline plugins/transport-nio/src/main/java/org/elasticsearch/http/nio/HttpReadWriteHandler.java:149]: I guess we can revert changes to this method now?

**Review by original-brownbear — APPROVED** [review]
LGTM thanks! Maybe that one noisy spot I commented on.

**Pull request merged**